### ISSUE TITLE
BYPASS/BUG Survivalfly setback allows you to avoid falling in the void

### ISSUE THREAD

**Issue by Alexre** [NoCheatPlus/Issues#516]
#### Short description of the issue.
When you are falling, if you right click the minecraft window, you will freeze and ncp detects you as you are using flight, making you move back to the block up there. (see the video to understand the process)
#### Server ####
git-TacoSpigot-"27784328" ~MC: 1.8.8~
  detected: 1.8.8
#### NoCheatPlus ####
Plugin: 3.15.2-SNAPSHOT-sMD5NET-b1113
MCAccess: 1.8.4-1.8.8 / Spigot-CB-1.8_R3
Features:
  blocks: BlocksMC1_4 | BlocksMC1_5 | BlocksMC1_6_1 | BlocksMC1_7_2 | BlocksMC1_8
  checks: FastConsume | Gutenberg | AttackFrequency | FlyingFrequency | KeepAliveFrequency
  defaults: pvpKnockBackVelocity
  packet-listeners: UseEntityAdapter | MovingFlying | OutgoingPosition | KeepAliveAdapter | SoundDistance
Hooks: AllViolations~NCP~ 1.0
#### Related Plugins ####
ProtocolLib v4.3.0 | ViaVersion v1.2.0
#### How to reproduce the issue (is it random / always happens / side conditions)?
See the video to know how to reproduce the issue: https://www.youtube.com/watch?v=rx9TqoAQ9lM
The log of NCP doing this action: https://hastebin.com/azohosopec.pas 
#### Have you made any (noteworthy) changes to the default configuration of NCP?
Using https://hastebin.com/magejobubo.swift - However, I happens with the default config of ncp too.
#### Do you run any other special plugins that affect game mechanics next to NC+ (adding skills, machines, adding/changing blocks, other anti cheating)?
No
#### Provide additional information, such as error/debug/violation logs. (For reproducible issues with moving consider: https://github.com/NoCheatPlus/Docs/wiki/Debugging#on-the-fly-debug-output-for-individual-players) Use GitHub gists or http://pastebin.com/ or similar for huge logs, or attach a (zip-) file here. For image uploads, please prefer a site that allows viewing without JavaScript, you can upload file attachments for issues right here.
https://hastebin.com/riropomumu.sql (logs from a default config, when the player does the action)
https://www.youtube.com/watch?v=rx9TqoAQ9lM

**Comment by Janmm14** [comment]
Duplicate of #8

**Comment by asofold** [comment]
Indeed a duplicate:
* The classic set-back policy applied by NCP has always been to set back to the last ground position, which isn't optimal for skywars like setups. The point is that you can otherwise reach places that you could/should not reach, examples:
    * Setting back into air: speeds don't match and places may become reachable/doable (canyon + window).
    * Setting down to ground. While this is not a bad candidate to add as an option for falling, it might allow too short moves as well, thus at least in theory (not considering relevance), this also is the category of enabling reaching places.
* Apart from that, clients could implement a 'never fall down' thing in more or less smart ways, and the discussion about where to set back for falling is gone for most. You can try to battle it for special cases like receiving velocity and modeling friction 100% accurate, but they still have ways of delaying getting pushed, due to the nature of the networking connections. So all in all, this is really nice to have and better should be implemented for skywars, but it's not really preventing cheating in any way.

A voidtovoid option already exists: once the player reaches the void, they are not teleported back. By default it's activated (checks.moving.survivalfly.setbackpolicy.voidtovoid). This just doesn't cover the part from an island/ground down to the void.

Concerning pulling players down to ground for falling, it'll certainly be implemented, but not before i've consoldiated the current change set.

**Comment by Alexre** [comment]
@asofold I already have the voidtovoid option enabled and the thing still happens

**Comment by asofold** [comment]
@Alexre 

* Do they get teleported out of the void itself or from falling towards the void?
* Could you post violation log messages (best with default configuration)?

**Comment by Alexre** [comment]
@asofold You are teleported to the land back, even if this option is enabled.
The video was recorded with the default config, I had the option enabled with the custom one too. 
So the log could be the same https://hastebin.com/riropomumu.sql

**Comment by asofold** [comment]
Teleported back: yes, it's intended behavior of NCP, _unless you are already inside the void_ (y < 0.0). 

Edit: The video shows you taking damage in the void and getting teleported back ?

Edit2: The hover violations should happen after 4 seconds of _not moving_ (roughly), the video looks like straight falling and sudden set back. That's odd.

**Comment by asofold** [comment]
The video shows you taking damage in the void and getting teleported back ?

**Comment by asofold** [comment]
Alright, i did not read the initial message fully - so the side conditions are clear:
* Make the client fail the hover check, using a vanilla client.
* This happens clearly _above_ the void at y coordinates around 30.

This means the voidtovoid option doesn't take effect, because you're not actually insode the void. So this is expected behavior of NCP, setting you back to the last ground position.

I might consider adding something simple for maps that are empty below the player, to start off simple. Not sure i'll do it before finishing the current hot topic, though. (Hopefully  weekend, not 100% sure.)

**Comment by Alexre** [comment]
@asofold I wasn't considering the void < 0 sorry, it happens > 0

**Comment by asofold** [comment]
I hadn't read the basic information at all :p. Excuse could be that i'm internally busy with the data storage overhaul, but still.

**Comment by Alexre** [comment]
@asofold So there is not any way to stop the setback when players are falling unless they are y < 0 ?

**Comment by asofold** [comment]
Currently not. Something can be implemented, but there are a range of pitfalls/questions with alternative set back policies. I'll certainly consider something for skywars like situations.

**Comment by asofold** [comment]
An experimental version will be coming with the next build - however it'll have to wait until the other changes i'm focused on have been implemented.

**Comment by asofold** [comment]
Supposedly fixed in the EXPERIMENTAL build 1144.

**Comment by Alexre** [comment]
After this, users can be teleported to the void while doing bridges

They made a video: https://youtu.be/B-1mQwzJhys 0:40

**Comment by asofold** [comment]
@Alexre 
Yes, it looks like i forgot or deliberately left out side condition checks (optional or not), for when to drag down.

Opened an extra issue for this: https://github.com/NoCheatPlus/Issues/issues/553
